core/search.go

```diff
@@ -0,0 +1,224 @@
+package core
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"os/exec"
+	"strings"
+
+	"dagger.io/dagger/telemetry"
+	"github.com/dagger/dagger/dagql"
+	"github.com/dagger/dagger/engine/slog"
+	"github.com/vektah/gqlparser/v2/ast"
+)
+
+type SearchResult struct {
+	FilePath       string            `field:"true" doc:"The path to the file that matched."`
+	LineNumber     int               `field:"true" doc:"The first line that matched."`
+	AbsoluteOffset int               `field:"true" doc:"The byte offset of this line within the file."`
+	MatchedLines   string            `field:"true" doc:"The line content that matched."`
+	Submatches     []*SearchSubmatch `field:"true" doc:"Sub-match positions and content within the matched lines."`
+}
+
+func (*SearchResult) Type() *ast.Type {
+	return &ast.Type{
+		NamedType: "SearchResult",
+		NonNull:   true,
+	}
+}
+
+type SearchSubmatch struct {
+	Text  string `field:"true" doc:"The matched text."`
+	Start int    `field:"true" doc:"The match's start offset within the matched lines."`
+	End   int    `field:"true" doc:"The match's end offset within the matched lines."`
+}
+
+func (*SearchSubmatch) Type() *ast.Type {
+	return &ast.Type{
+		NamedType: "SearchSubmatch",
+		NonNull:   true,
+	}
+}
+
+type SearchOpts struct {
+	Pattern    string
+	Literal    bool `default:"false"`
+	Multiline  bool `default:"false"`
+	Dotall     bool `default:"false"`
+	IgnoreCase bool `default:"false"`
+	FilesOnly  bool `default:"false"`
+	Limit      *int
+}
+
+func (opts SearchOpts) Args() []dagql.Argument {
+	return []dagql.Argument{
+		dagql.Arg("pattern").Doc(`The text to match.`),
+		dagql.Arg("literal").Doc(`Interpret the pattern as a literal string instead of a regular expression.`),
+		dagql.Arg("multiline").Doc(`Enable searching across multiple lines.`),
+		dagql.Arg("dotall").Doc(`Allow the . pattern to match newlines in multiline mode.`),
+		dagql.Arg("ignoreCase").Doc(`Enable case-insensitive matching.`),
+		dagql.Arg("filesOnly").Doc(`Only return matching files, not lines and content`),
+		dagql.Arg("limit").Doc(`Limit the number of results to return`),
+	}
+}
+
+func (opts SearchOpts) RipgrepArgs() []string {
```

Comment:
IMO we should add ` --no-ignore` so it doesn't honor `.gitignore` by default and `--hidden` so it doesn't skip things that start with `.`. I can definitely see those being useful, but they feel better as options in our API (probably defaulting off).

Reply:
Done! Added `skipIgnore` and `skipHidden` and renamed `ignoreCase` to `insensitive` so it's not confused with ignore-files